File: core/fallDetectorCore.ts
// core/fallDetectorCore.ts
export type FallThresholds = {
  rateMs: number;
  impactG: number;
  stillnessG: number;
  impactGraceMs: number;
  stillnessWindowMs: number;
  stillnessMinMs: number;
  cooldownMs: number;
  maxObserveMs: number;
};

type Sample = { t: number; g: number };

const DEFAULTS: FallThresholds = {
  rateMs: 50,                 // ~20 Hz
  impactG: 8.0,               // hard impact
  stillnessG: 1.05,           // “still” on devices that idle ~1g
  impactGraceMs: 1200,        // let post-impact jiggle die down
  stillnessWindowMs: 1200,    // sliding window size
  stillnessMinMs: 700,        // consecutive ms under threshold
  cooldownMs: 10_000,         // one alert max per 10s
  maxObserveMs: 4000,         // how long after grace to look for stillness
};

export type FallDetector = {
  onSample: (g: number, now?: number) => void;
  reset: () => void;
  setThresholds: (patch: Partial<FallThresholds>) => void;
  getConfig: () => FallThresholds;
  getState: () => { state: "IDLE" | "AFTER_IMPACT"; lastTrigger: number; impactAt: number };
};

export function createFallDetector(
  onFall: () => void,
  opts?: { thresholds?: Partial<FallThresholds>; debug?: boolean }
): FallDetector {
  let cfg: FallThresholds = { ...DEFAULTS, ...(opts?.thresholds ?? {}) };
  let state: "IDLE" | "AFTER_IMPACT" = "IDLE";
  let impactAt = 0;
  let lastTrigger = 0;
  let buf: Sample[] = [];

  const log = (...a: any[]) => opts?.debug && console.log("[fall]", ...a);

  const prune = (now: number) => {
    const keepSince = now - Math.max(cfg.impactGraceMs + cfg.maxObserveMs + cfg.stillnessWindowMs + 1000, cfg.cooldownMs);
    while (buf.length && buf[0].t < keepSince) buf.shift();
  };

  const onSample = (g: number, now = Date.now()) => {
    // push & prune
    buf.push({ t: now, g });
    prune(now);

    // cooldown gate
    if (now - lastTrigger < cfg.cooldownMs) return;

    if (state === "IDLE") {
      if (g >= cfg.impactG) {
        state = "AFTER_IMPACT";
        impactAt = now;
        log("IMPACT", { g: g.toFixed(2) });
      }
      return;
    }

    // AFTER_IMPACT
    const dt = now - impactAt;
    if (dt < cfg.impactGraceMs) return;

    // sliding stillness window up to 'now'
    const slideStart = now - cfg.stillnessWindowMs;
    let sum = 0, n = 0, stillMs = 0;

    for (let i = 1; i < buf.length; i++) {
      const prev = buf[i - 1], cur = buf[i];
      if (cur.t < slideStart) continue;
      sum += cur.g; n++;
      if (prev.t >= slideStart && prev.g <= cfg.stillnessG && cur.g <= cfg.stillnessG) {
        stillMs += cur.t - prev.t;
      }
    }

    const avg = n ? sum / n : Number.POSITIVE_INFINITY;
    log("ANALYZE", `avg=${Number.isFinite(avg) ? avg.toFixed(2) : "inf"}`, `stillMs=${stillMs}ms`, `dt=${dt}ms`);

    if (avg <= cfg.stillnessG || stillMs >= cfg.stillnessMinMs) {
      lastTrigger = now;        // starts cooldown
      state = "IDLE";
      log("FALL DETECTED ✅");
      onFall();
      buf = buf.filter(s => now - s.t < 300); // trim to avoid immediate retrigger
      return;
    }

    // give up if no stillness appears
    if (dt > cfg.impactGraceMs + cfg.maxObserveMs) {
      state = "IDLE";
      log("RESET (no stillness)");
    }
  };

  const reset = () => { state = "IDLE"; impactAt = 0; lastTrigger = 0; buf = []; };
  const setThresholds = (patch: Partial<FallThresholds>) => { cfg = { ...cfg, ...patch }; };
  const getConfig = () => cfg;
  const getState = () => ({ state, lastTrigger, impactAt });

  return { onSample, reset, setThresholds, getConfig, getState };
}

File: android/app/src/main/java/com/miguell0706/FallNotifier3/FallDetectorCore.kt
package com.miguell0706.FallNotifier3

import android.util.Log
import kotlin.math.max

// --- Threshold config (like FallThresholds in TS) ---

data class FallThresholds(
    val rateMs: Int = 50,                // ~20 Hz
    val impactG: Float = 8.0f,           // hard impact
    val stillnessG: Float = 1.05f,       // “still” idle ~1g
    val impactGraceMs: Long = 1200L,     // let post-impact jiggle die down
    val stillnessWindowMs: Long = 1200L, // sliding window size
    val stillnessMinMs: Long = 700L,     // consecutive ms under threshold
    val cooldownMs: Long = 10_000L,      // one alert max per 10s
    val maxObserveMs: Long = 4000L       // how long after grace to look for stillness
)

// Simple sample type: time + g
private data class Sample(val t: Long, val g: Float)

// --- State info (like getState TS return) ---

enum class FallDetectorPhase { IDLE, AFTER_IMPACT }

data class FallDetectorState(
    val state: FallDetectorPhase,
    val lastTrigger: Long,
    val impactAt: Long
)

// --- Interface (to match what FallEngine expects) ---

interface FallDetector {
    fun onSample(g: Float, now: Long = System.currentTimeMillis())
    fun reset()
    fun setThresholds(patch: FallThresholdsPatch)
    fun getConfig(): FallThresholds
    fun getState(): FallDetectorState
}

// Patch type, equivalent of Partial<FallThresholds> in TS
data class FallThresholdsPatch(
    val rateMs: Int? = null,
    val impactG: Float? = null,
    val stillnessG: Float? = null,
    val impactGraceMs: Long? = null,
    val stillnessWindowMs: Long? = null,
    val stillnessMinMs: Long? = null,
    val cooldownMs: Long? = null,
    val maxObserveMs: Long? = null
)

// --- Implementation (this is the TS createFallDetector port) ---

class FallDetectorCore(
    private val onFall: () -> Unit,
    thresholds: FallThresholds = FallThresholds(),
    private val debug: Boolean = false
) : FallDetector {

    private var cfg: FallThresholds = thresholds
    private var phase: FallDetectorPhase = FallDetectorPhase.IDLE
    private var impactAt: Long = 0L
    private var lastTrigger: Long = 0L
    private val buf = mutableListOf<Sample>()

    private fun log(vararg parts: String) {
        if (!debug) return
        Log.d("FallDetector", parts.joinToString(" "))
    }

    private fun prune(now: Long) {
        // keepSince = now - max(impactGraceMs + maxObserveMs + stillnessWindowMs + 1000, cooldownMs)
        val span = cfg.impactGraceMs + cfg.maxObserveMs + cfg.stillnessWindowMs + 1000L
        val keepSince = now - max(span, cfg.cooldownMs)
        while (buf.isNotEmpty() && buf.first().t < keepSince) {
            buf.removeAt(0)
        }
    }

    override fun onSample(g: Float, now: Long) {
        // push & prune
        buf.add(Sample(now, g))
        prune(now)

        // cooldown gate
        if (now - lastTrigger < cfg.cooldownMs) return

        if (phase == FallDetectorPhase.IDLE) {
            if (g >= cfg.impactG) {
                phase = FallDetectorPhase.AFTER_IMPACT
                impactAt = now
                log("IMPACT g=${"%.2f".format(g.toDouble())}")
            }
            return
        }

        // AFTER_IMPACT
        val dt = now - impactAt
        if (dt < cfg.impactGraceMs) return

        // sliding stillness window up to 'now'
        val slideStart = now - cfg.stillnessWindowMs
        var sum = 0f
        var n = 0
        var stillMs = 0L

        for (i in 1 until buf.size) {
            val prev = buf[i - 1]
            val cur = buf[i]

            if (cur.t < slideStart) continue

            sum += cur.g
            n++

            if (prev.t >= slideStart &&
                prev.g <= cfg.stillnessG &&
                cur.g <= cfg.stillnessG
            ) {
                stillMs += (cur.t - prev.t)
            }
        }

        val avg: Float =
            if (n > 0) sum / n.toFloat() else Float.POSITIVE_INFINITY
        val avgStr = if (avg.isFinite()) "%.2f".format(avg.toDouble()) else "inf"

        log("ANALYZE avg=$avgStr stillMs=${stillMs}ms dt=${dt}ms")

        // Condition for fall:
        // avg <= stillnessG OR stillMs >= stillnessMinMs
        if (avg <= cfg.stillnessG || stillMs >= cfg.stillnessMinMs) {
            lastTrigger = now
            phase = FallDetectorPhase.IDLE
            log("FALL DETECTED ✅")
            onFall()
            // trim to avoid immediate retrigger
            val keepRecentSince = now - 300L
            val keep = buf.filter { s -> now - s.t < 300L }
            buf.clear()
            buf.addAll(keep)
            return
        }

        // give up if no stillness appears
        if (dt > cfg.impactGraceMs + cfg.maxObserveMs) {
            phase = FallDetectorPhase.IDLE
            log("RESET (no stillness)")
        }
    }

    override fun reset() {
        phase = FallDetectorPhase.IDLE
        impactAt = 0L
        lastTrigger = 0L
        buf.clear()
    }

    override fun setThresholds(patch: FallThresholdsPatch) {
        cfg = cfg.copy(
            rateMs = patch.rateMs ?: cfg.rateMs,
            impactG = patch.impactG ?: cfg.impactG,
            stillnessG = patch.stillnessG ?: cfg.stillnessG,
            impactGraceMs = patch.impactGraceMs ?: cfg.impactGraceMs,
            stillnessWindowMs = patch.stillnessWindowMs ?: cfg.stillnessWindowMs,
            stillnessMinMs = patch.stillnessMinMs ?: cfg.stillnessMinMs,
            cooldownMs = patch.cooldownMs ?: cfg.cooldownMs,
            maxObserveMs = patch.maxObserveMs ?: cfg.maxObserveMs
        )
    }

    override fun getConfig(): FallThresholds = cfg

    override fun getState(): FallDetectorState =
        FallDetectorState(phase, lastTrigger, impactAt)
}
